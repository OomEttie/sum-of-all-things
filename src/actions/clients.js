const uuidV4 = require('uuid/v4');
import database, { clientsDatabaseRef } from '../utils/firebase';

import {
  ACTION_CLIENT_ADD,
  ACTION_CLIENT_EDIT,
  ACTION_CLIENT_LIST,
  ACTION_CLIENT_REMOVE
} from '../actions/actions-list';

//
// ACTION_CLIENT_ADD
//
export const addClient = client => ({
  type: ACTION_CLIENT_ADD,
  client
});

export const startAddClient = (clientData = {}) => {
  return (dispatch, getState) => {
    const { name = '', surname = '', note = '' } = clientData;

    const client = { name, surname, note };

    const dbPath = clientsDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );

    return database
      .ref(dbPath)
      .push(client)
      .then(ref => {
        dispatch(
          addClient({
            id: ref.key,
            ...client
          })
        );
      });
  };
};

//
// REMOVE_EXPENSE
//
export const removeClient = ({ id } = {}) => ({
  type: ACTION_CLIENT_REMOVE,
  id: id
});

export const startRemoveClient = ({ id }) => {
  return (dispatch, getState) => {
    const dbPath = clientsDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );

    return database
      .ref(`${dbPath}/${id}`)
      .remove()
      .then(data => {
        dispatch(removeClient({ id }));
      });
  };
};

//
// EDIT_CLIENT
//
export const editClient = (id, updates) => ({
  type: ACTION_CLIENT_EDIT,
  id,
  updates
});

export const startEditClient = (id, updates) => {
  return (dispatch, getState) => {
    let dbPath = clientsDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );

    return database
      .ref(`${dbPath}/${id}`)
      .update({
        ...updates
      })
      .then(() => {
        dispatch(editClient(id, updates));
      });
  };
};

//
// LIST CLIENTS
//
export const listClients = clients => ({
  type: ACTION_CLIENT_LIST,
  clients
});

export const startListClients = () => {
  return (dispatch, getState) => {
    const dbPath = clientsDatabaseRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(dbPath)
      .once('value')
      .then(snapshot => {
        const clientsData = [];
        snapshot.forEach(snapshotChild => {
          clientsData.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        dispatch(listClients(clientsData));
      })
      .catch(e => {});
  };
};
