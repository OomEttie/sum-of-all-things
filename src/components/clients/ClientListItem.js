import React from 'react';
import { Link } from 'react-router-dom';

export const ClientListItem = ({ dispatch, id, name, surname, note }) => (
  <div className="button-material-primary">
    <Link className="list-item" to={`/clients/edit/${id}`}>
      <div>
        <h3 className="list-item__title">
          {name} {surname}
        </h3>
        <span className="list-item__subtitle">{note}</span>
      </div>
    </Link>
  </div>
);

export default ClientListItem;
