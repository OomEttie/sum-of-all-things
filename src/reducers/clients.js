export default (state = {}, action) => {
  switch (action.type) {
    case 'ADD_EXPENS':
      return {
        uid: action.uid
      };
    case 'LOGOUT':
      console.log('LOGOUT');
      return {};
    default:
      return state;
  }
};
