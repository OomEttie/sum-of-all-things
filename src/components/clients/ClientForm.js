import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';

import { clientObject } from '../../utils/__mocks__/clients';

export class ClientForm extends React.Component {
  constructor(props) {
    super(props);

    const clientData = clientObject();
    clientData.id = props.client ? props.client.id : '';
    clientData.name = props.client ? props.client.name : '';
    clientData.surname = props.client ? props.client.surname : '';
    clientData.note = props.client ? props.client.note : '';

    this.state = {
      ...clientData,
      error: '',
      open: false
    };
  }

  onNameChange = e => {
    const name = e.target.value;
    this.setState(prevState => ({ name: name }));
  };

  onSurnameChange = e => {
    const surname = e.target.value;
    this.setState(prevState => ({ surname: surname }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(prevState => ({ note: note }));
  };

  onRemoveClient = e => {
    this.props.onRemoveClient(this.state.id);
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.name || !this.state.surname) {
      this.setState(prevState => ({
        error: 'Please provide name and surname!'
      }));
    } else {
      this.setState(() => ({
        error: ''
      }));
      this.props.onSubmit({
        name: this.state.name,
        surname: this.state.surname,
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error.length > 0 && (
            <p className="form__error">{this.state.error}</p>
          )}
          <input
            className="text-input"
            type="text"
            placeholder="Name"
            autoFocus
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <input
            className="text-input"
            type="text"
            placeholder="Surname"
            value={this.state.surname}
            onChange={this.onSurnameChange}
          />
          <textarea
            className="textarea"
            placeholder="Add a note (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <div>
            <button>{this.props.client ? 'Save' : 'Add'} Client</button>
          </div>
        </form>
        {this.props.client && (
          <button onClick={this.onRemoveClient}>Remove Client</button>
        )}
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {};
};

export default connect(mapStateToProps)(ClientForm);
