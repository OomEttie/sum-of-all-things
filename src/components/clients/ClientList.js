import React from 'react';
import { connect } from 'react-redux';
import { ClientListItem } from './ClientListItem';

import { Link } from 'react-router-dom';

export const ClientList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="header__content">
        <div className="show-for-mobile">Clients MOB</div>
        <div className="show-for-desktop">Clients</div>
      </div>
      <Link className="button " to="/clients/add">
        Add Client
      </Link>
    </div>
    <div className="list-body">
      {props.clients.length == 0 ? (
        <div className="list-item list-item--message">
          <span>No Clients</span>
        </div>
      ) : (
        props.clients.map(client => {
          return <ClientListItem key={client.id} {...client} />;
        })
      )}
    </div>
  </div>
);
ClientList.defaultProps = {
  clients: []
};

export default ClientList;
