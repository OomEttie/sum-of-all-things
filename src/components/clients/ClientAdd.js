import React from 'react';
import { connect } from 'react-redux';
import ClientForm from './ClientForm';
import { startAddClient } from '../../actions/clients';
import { clients } from '../../utils/__mocks__/clients';

export class ClientAdd extends React.Component {
  onSubmit = client => {
    this.props.addClient(client);
    this.props.history.push('/clients');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Client</h1>
          </div>
        </div>
        <div className="content-container">
          <ClientForm onSubmit={this.onSubmit}/>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addClient: client => {
      dispatch(startAddClient(client));
    }
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(ClientAdd);
