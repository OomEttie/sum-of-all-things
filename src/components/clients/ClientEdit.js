import React from 'react';
import { connect } from 'react-redux';
import ClientForm from './ClientForm';
import { startEditClient, startRemoveClient } from '../../actions/clients';

export class ClientEdit extends React.Component {
  onSubmit = client => {
    this.props.editClient(this.props.client.id, { ...client });
    this.props.history.push('/clients');
  };

  onRemoveClient = id => {
    this.props.removeClient(id);
    this.props.history.push('/clients');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Client</h1>
          </div>
        </div>
        <div className="content-container">
          <ClientForm
            client={this.props.client}
            onSubmit={this.onSubmit}
            onRemoveClient={this.onRemoveClient}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    client: state.clients.find(client => {
      return client.id == props.match.params.id;
    })
  };
};

const mapDispatchToProps = dispatch => {
  return {
    editClient: (id, updates) => {
      dispatch(startEditClient(id, updates));
    },
    removeClient: id => {
      dispatch(startRemoveClient({ id: id }));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ClientEdit);
