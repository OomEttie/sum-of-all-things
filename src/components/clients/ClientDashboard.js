import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { ClientList } from './ClientList';
import { startListClients } from '../../actions/clients';

export class ClientDashboard extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div className="content-container">
        <ClientList clients={this.props.clients} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    clients: state.clients
  };
};

const mapDispatchToProps = dispatch => {};

export default connect(mapStateToProps)(ClientDashboard);
