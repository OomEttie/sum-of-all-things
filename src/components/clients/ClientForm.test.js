import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';

import ClientForm from './ClientForm';
import { clients } from '../../utils/__mocks__/clients';

test('render client form correctly', () => {
  const wrapper = shallow(<ClientForm />);
  expect(wrapper).toMatchSnapshot();
});

test('render client form with data', () => {
  const wrapper = shallow(<ClientForm client={clients[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('render client form with invalid data', () => {
  const wrapper = shallow(<ClientForm />);
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('call onSubmit for valid submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ClientForm client={clients[1]} onSubmit={onSubmitSpy} />
  );
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });

  expect(wrapper.state('error').length).toBe(0);
  expect(onSubmitSpy).toHaveBeenCalled();
  expect(onSubmitSpy).toHaveBeenCalledWith({
    name: clients[1].name,
    surname: clients[1].surname,
    note: clients[1].note
  });
});
