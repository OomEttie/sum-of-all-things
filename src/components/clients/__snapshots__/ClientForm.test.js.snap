// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render client form correctly 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      className="text-input"
      onChange={[Function]}
      placeholder="Surname"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Add a note (optional)"
      value=""
    />
    <div>
      <button>
        Add
         Client
      </button>
    </div>
  </form>
</div>
`;

exports[`render client form with data 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="2 name"
    />
    <input
      className="text-input"
      onChange={[Function]}
      placeholder="Surname"
      type="text"
      value="2 surname"
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Add a note (optional)"
      value="2 note"
    />
    <div>
      <button>
        Save
         Client
      </button>
    </div>
  </form>
  <button
    onClick={[Function]}
  >
    Remove Client
  </button>
</div>
`;

exports[`render client form with invalid data 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <p
      className="form__error"
    >
      Please provide name and surname!
    </p>
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      className="text-input"
      onChange={[Function]}
      placeholder="Surname"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Add a note (optional)"
      value=""
    />
    <div>
      <button>
        Add
         Client
      </button>
    </div>
  </form>
</div>
`;
